0006-bingings.md

```diff
@@ -0,0 +1,591 @@
+* Currently, we also pass the length together with `header_hash_ptr` and `mix_hash_ptr` which is not necessary since
+we know their length.
+
+## Promises API
+
+```rust
+promise_create(account_id_len: u64,
+               account_id_ptr: u64,
+               method_name_len: u64,
+               method_name_ptr: u64,
+               arguments_len: u64,
+               arguments_ptr: u64,
+               amount_ptr: u64) -> u64
+```
+Creates a promise that will execute a method on account with given arguments and attaches the given amount.
+`amount_ptr` points to a slice of bytes representing `u128`.
+
+###### Panics
+* If `account_id_len + account_id_ptr` or `method_name_len + method_name_ptr` or `arguments_len + arguments_ptr`
+or `amount_ptr + 8` points outside the memory of the guest or host, with `MemoryAccessViolation`.
+
+###### Returns
+* Index of the new promise that uniquely identifies it within the current execution of the method.
+
+---
+
+```rust
+promise_then(promise_idx: u64,
+             account_id_len: u64,
+             account_id_ptr: u64,
+             method_name_len: u64,
+             method_name_ptr: u64,
+             arguments_len: u64,
+             arguments_ptr: u64,
+             amount_ptr: u64) -> u64            
+```
+Attaches the callback that is executed after promise pointed by `promise_idx` is complete.
+
+###### Panics
+* If `promise_idx` does not correspond to an existing promise panics with `InvalidPromiseIndex`.
+* If `account_id_len + account_id_ptr` or `method_name_len + method_name_ptr` or `arguments_len + arguments_ptr`
+or `amount_ptr + 8` points outside the memory of the guest or host, with `MemoryAccessViolation`.
+
+###### Returns
+* Index of the new promise that uniquely identifies it within the current execution of the method.
+
+---
+```rust
+promise_and(promise_id1: u64, promise_id2: u64) -> u64
+```
+Creates a new promise which completes at the same time both promises complete.
+
+###### Panics
+* If `promise_idx1` or `promise_idx2` do not correspond to existing promises panics with `InvalidPromiseIndex`.
+
+###### Returns
+* Index of the new promise that uniquely identifies it within the current execution of the method.
+
+---
+```rust
+promise_results_count() -> u64
+```
+If the current function is invoked by a callback we can access the execution results of the promises that
+caused the callback. This function returns the number of complete and incomplete callbacks.
+
+Note, we are only going to have incomplete callbacks once we have `promise_or` combinator.
+###### Normal execution
+* If there is only one callback `promise_results_count()` returns `1`;
+* If there are multiple callbacks (e.g. created through `promise_and`) `promise_results_count()` returns their number.
+* If the function was called not through the callback `promise_results_count()` returns `0`.
+
+
+---
+```rust
+promise_result(promise_idx: u64, register_id: u64)
```

Comment:
this is where we need to distinguish `promise_idx` and `result_index`. In this case this is a result index and it's only available within callbacks. So the actual promise_idx would be different, unless you suggest to inject it there. Might be an interesting concept, but right now positional results are enough.

Reply:
Got it.